Add support for one-click onboarding

Gudahtt opened the pull request:
MetaMask now allows sites to register as onboarding the user, so that the user is redirected back to the initiating site after onboarding. This is accomplished through the use of the `metamask-onboarding` library and the MetaMask forwarder.

At the end of onboarding, the text of the button will show the origin of the site the user is about to be redirected to. This is intended to help prevent phishing attempts, as it highlights that a redirect is taking place to an untrusted third party.

The onboarding library has been demonstrated with the `contract-test` dapp used for some of our e2e tests. That dapp now uses a "Connect" button instead of connecting automatically, and acts as an onboarding button if MetaMask isn't detected.

A new static server was written to allow referencing the `metamask-onboarding` library in the dapp without vendoring it. This server replaces `static-server`, which wasn't able to accomplish this due to a bug in how symlinks are handled.

This PR depends upon [this update to the forwarder](https://github.com/MetaMask/forwarder/pull/2), and this [new onboarding library](https://github.com/Gudahtt/metamask-onboarding/pull/1).

Here is a diagram of the interactions between the onboarding library, the forwarder, and the extension:
![Onboarding Flow](https://user-images.githubusercontent.com/2459287/67541693-439c9600-f6c0-11e9-93f8-112a8941384a.png)

Here is the onboarding flow with a new wallet (using the 'Remind me later' option):
![onboarding_incremental](https://user-images.githubusercontent.com/2459287/66891257-08fb7500-efbf-11e9-8e8e-690667680ff7.gif)

Here is the onboarding flow when importing a wallet: 
![onboarding_import](https://user-images.githubusercontent.com/2459287/66891283-19abeb00-efbf-11e9-9cd5-30a15615cffc.gif)

(These two recordings are using an older test dapp that I wrote and have since discarded - the `contract-test` dapp should work just as well for testing)

This can be tested locally by running the `contract-test` dapp (`yarn dapp`), and also running the forwarder locally on port 9010. To run the forwarder, cloning my forwarder fork ([the `forwarder-rewrite` branch](https://github.com/Gudahtt/forwarder/tree/forwarder-rewrite)), install dependencies with `yarn`, then starting the forwarder with `yarn start`. Then the flow shown above should work.

Gudahtt committed:
Add support for one-click onboarding

MetaMask now allows sites to register as onboarding the user, so that
the user is redirected back to the initiating site after onboarding.
This is accomplished through the use of the `metamask-onboarding`
library and the MetaMask forwarder.

At the end of onboarding, the text of the button will show the origin
of the site the user is about to be redirected to. This is intended to
help prevent phishing attempts, as it highlights that a redirect is
taking place to an untrusted third party.

Closes #6161

Gudahtt committed:
Add onboarding button to contract test dapp

The `contract-test` dapp (run with `yarn dapp`, used in e2e tests) now
uses a `Connect` button instead of connecting automatically. This
button also serves as an onboarding button when a MetaMask installation
is not detected.

Gudahtt committed:
Add new static server for test dapp

The `static-server` library we were using for the `contract-test` dapp
didn't allow referencing files outside the server root. This should
have been possible to work around using symlinks, but there was a bug
that resulted in symlinks crashing the server.

Instead it has been replaced with a simple static file server that
will serve paths starting with `node_modules` from the project root.
This will be useful in testing the onboarding library without vendoring
it.

danjm commented on app/scripts/background.js:
```diff
@@ -313,6 +313,7 @@ function setupController (initState, initLangCode) {
   //
   extension.runtime.onConnect.addListener(connectRemote)
   extension.runtime.onConnectExternal.addListener(connectExternal)
+  extension.runtime.onMessage.addListener(controller.onMessage.bind(controller))
```
Does this only listen for events emitted by explicit calls to `extension.runtime.sendMessage()`?

Gudahtt replied:
Yes, I believe so. Either that or `extension.tabs.sendMessage`, according to [this](https://developer.chrome.com/extensions/runtime#event-onMessage) and [this](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage).

danjm commented:
So I read through all the code of all three PRs. It all made sense to me, but I felt that I am not reading carefully enough as I have no real comments or suggestions for improvement yet.

I went back through more carefully, took more detailed notes of what is happening, and came up with  a few clarifying questions. So far, this more careful re-read has only covered the other two PRs (not this 7017 one). I believe I fully understand how they work now and assuming my questions are answered without issue, I think I can say that the interaction of the forwarder and the onboarder looks good.

Here are my notes on how the standard case works for the interaction of the forwarder and the onboarder:

```
  1. Dapp instantiates MetamaskOnboarding
    - during instantiation, a message listener is added which handles messages with the `_onMessage()` method
    - `this.state` will be set to `ONBOARDING_STATE.NOT_INSTALLED`
  2. Dapp adds a button to the dom that checks that MetaMask is not installed and calls `startOnboarding()` on the instance of MetaMaskOnboarding
  3. `startOnboarding()` does three things: sets `REGISTRATION_IN_PROGRESS` in session storage to true; calls `_openDownloadPage()`; calls `_openForwarder()`
    3a. `_openDownloadPage()` opens a tab pointed to the page on the chrome store where metamask can be downloaded
    3b. `_openForwarder()` injects an invisible iframe into the dapp page which has `'https://fwd.metamask.io'` as its source
    3c. The user is now looking at the chrome store and starts to manually go through the download/install + onboarding flow
  4. Meanwhile, the injected forwarder code (running in the dapp tab) instantiates a `Forwarder` and calls `initialize` once the dom content (of the iframe) has loaded
  5. `initialize()` checks if metamask is installed, and if it is not, it reloads the iframe which will result in another call to `initialize()`
    - note that the check for whether metamask is installed is based on whether the injected ethereum global has a truthy `isMetaMask` property
    - this reloading will happen every three seconds untill MetaMask has actually been installed, without the user knowing it is happening
  6. if, when `initialize is called, metamask is installed, three things will happen
    - the forwarder's `this.parentWindow` property will be pointed to the dapp page window
    - A listener for messages will be added to the iframe window
    - an interval will be created (`this.reloadParentInterval`) that will post a message to the dapp window every second with data that has a `type` property equal to `'metamask:reload'`
  7. when the dapp window receives a message with data that has a type property equal to `'metamask:reload'`, the `_onMessageFromForwarder()` method is called
  8. when receiving the first `'metamask:reload'` message, `this.state` will be set to `ONBOARDING_STATE.RELOADING` and then the window will reload
  9. the dapp will reload and then instantiate MetamaskOnboarding again
    - this time, `this.state` will be set to ONBOARDING_STATE.INSTALLED
  10. when receiving the next `'metamask:reload'` message,
    - `this.state` will be set to `ONBOARDING_STATE.REGISTERING`,
    - `Onboarding._register()` will be called, which calls metamask's `registerOnboarding`
    - a message will be posted back to the forwarder that contains data with a `type` set to `'metamask:setParentLocation'` and a location property set to the location.href of the dapp window
  11. upon receiving a message from its parent window with a data object with a type of `'metamask:setParentLocation'`, the forwarder will
    - call `registerForwarder`, which calls metamask's `registerOnboardingForwarder` method with the location.href of the dapp (i.e. forwarder's parent) window
    - stop the interval that is sending the `'metamask:reload'` messages
    - post a `'metamask:registrationCompleted'` message to the parent (i.e. dapp) window
  12. when the dapp window receives a message with data that has a type property equal to `'metamask:registrationCompleted'`, the onboarding `stopOnboarding()` method is called, which will
    - call `_removeForwarder()` , which will remove the iframe
    - set `REGISTRATION_IN_PROGRESS` in session storage to false
  13. metamask's `registerOnboarding()` and `registerOnboardingForwarder()` methods each write an object to the onboardingStream that is create when the inpage script is run
    - `registerOnboarding()` writes an object with a `'registerOnboarding'` `type`
    - `registerOnboardingForwarder()` writes an object with a `'registerOnboardingForwarder'` `type`, and an `openerLocation` property set to the location.href of the dapp window
```

I am now going through a careful reread of the code changes in this PR and writing such notes. Will update with another comment in a bit.

Gudahtt commented on app/scripts/controllers/onboarding.js:
```diff
@@ -38,6 +59,43 @@ class OnboardingController {
+  }
+
+  /**
+   * Registering a site as having an onboarding button
```
This was a mistake; the description should be something more like:

>Registering a site as having initiated onboarding

Gudahtt replied:
This description needs updating as well

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1506,6 +1508,50 @@ module.exports = class MetamaskController extends EventEmitter {
+        handleMessage = this.onboardingController.registerOnboarding(message.location, sender.tab.id)
+      } else if (message.type === 'metamask:registerOnboardingForwarder') {
+        assert(message.openerLocation, 'Missing opener location')
+        // TODO: filter by forwarder domain unless in testing/development
```
This remains to be done. The forwarder needs to be deployed first for this to work though.

Gudahtt commented:
After considering this further, I think this can be simplified by removing the `registerOnboardingForwarder` step and moving the initial registration to after onboarding was triggered and MetaMask installed. I'll update the PR and and the diagram and such, though the videos should still be valid as this wouldn't be a user-facing change.

Open questions:

* Is the case where the dapp tab is closed handled adequately? Maybe we need to ensure the tab is still open, and redirect on the onboarding tab if not
* Do we want to allow the user to skip the redirect step?

Gudahtt committed:
Remove forwarder registration

The secondary forwarder registration step has been removed. It wasn't
useful anymore; it was originally intended to improve security when
one-click onboarding was architected differently, but it no longer
helps.

The onboarding state has been simpified to one set of onboarding
initiators, each with a `tabId`.

Gudahtt committed:
Rename 'origin' property to 'location'

This property includes the full URL for the onboarding initiator,
including the path.

Gudahtt committed:
Fallback to redirect if onboarding initiator is closed

If the onboarding initiator tab is closed when onboarding is finished,
the user is redirected to the onboarding originator as a fallback.

Gudahtt commented:
The forwarder registration has been removed from all three PRs. The diagram in the PR description has been updated as well.

I've also addressed the case where the initiating tab is no longer open. It'll now redirect in that case instead. This should be ready for review again now.

danjm reviewed (dismissed):
The forwarder updates look good and make sense. I am tentatively approving and will do one final look after design QA.

whymarrh reviewed (changes requested):
This LGTM! This is great work @Gudahtt!

I was able to build this branch and the forwarder branch and run both (`yarn dist && yarn dapp` and `yarn start`, respectively) and successfully see a link back to the dapp after onboarding. When nagivated to the Chrome Web Store to install MetaMask, I switched to `chrome://extensions` and loaded the unpacked extension.

<img width="1154" alt="" src="https://user-images.githubusercontent.com/1623628/68090420-200ae580-fe4e-11e9-9d83-596032d6275e.png">

This worked well at least twice through. I also think that the failure mode here is pretty good (that is, when this fails to link the dapp, it's no worse than the current on-boarding process).

I've left a few small comments but I would also like to politely block this until the following has been completed:

1. We update the forwarder and have that published (https://github.com/MetaMask/forwarder/pull/2)
2. We publish 1.x of the onboarding library (https://github.com/Gudahtt/metamask-onboarding/pull/1)
3. We update this PR and rebase it on the latest `develop`

I've left reviews on those PRs as well.

Gudahtt committed:
Use `loglevel` instead of `console`

Gudahtt commented:
Here are some screenshots of the changes to the button during the onboarding flow:

Here is the reveal seedphrase screen, which is where the "remind me later" button (e.g. incremental security) is shown:
![incremental_onboarding](https://user-images.githubusercontent.com/2459287/68248062-77e25180-fff2-11e9-8ee6-eb84fc69be8e.png)

This is the "onboarding complete" screen, which is shown after either importing a seedphrase, or confirming a new seedphrase:
![onboarding_complete](https://user-images.githubusercontent.com/2459287/68248070-7d3f9c00-fff2-11e9-8cc6-1155cdcf9987.png)

Here are a couple more screenshots of the reveal seedphrase screen with very long domain names. The latter is clearly broken, but the length is fairly absurd so I'm not sure how much we care.
![incremental_onboarding_long_domain](https://user-images.githubusercontent.com/2459287/68248244-cdb6f980-fff2-11e9-8d25-8345c9f518bf.png)
![incremental_onboarding_stupid_long_domain](https://user-images.githubusercontent.com/2459287/68248249-cee82680-fff2-11e9-9dd5-c7bddaf9fc78.png)

cc @cjeria

cjeria commented:
@Gudahtt after looking at this closer, I'm thinking we likely shouldn't add more text inside the buttons that could drastically change the appearance of our button styles, so I'd suggest we use a snackbar style notification just below the buttons. Check out these designs in [figma](https://www.figma.com/file/HEgBtmMHUQSlyQTyyeKSvD4R/One-click-Onboarding-Flow?node-id=660%3A3249) and screenshots see below:

![image](https://user-images.githubusercontent.com/597770/68256047-d6540380-ffe3-11e9-84ec-30dbed3dfa37.png)

Gudahtt commented:
That looks great! Good idea

Gudahtt committed:
Move return text to snackbar below button

The text explaining that the user was about to be moved back to the
dapp that initiated onboarding has been moved to a snackbar style
component below the button instead. This leaves more room to explain to
the user what to expect without making the button weirdly large.

Gudahtt commented:
Here are the updated buttons:

![reveal](https://user-images.githubusercontent.com/2459287/68355129-aa6c7700-00e4-11ea-9bac-fd55343a4383.png)
![congrats](https://user-images.githubusercontent.com/2459287/68355135-ad676780-00e4-11ea-97fa-de51a8e5fb58.png)

@cjeria does this look OK?

cjeria commented:
Looks good! @Gudahtt

Gudahtt committed:
Use updated `@metamask/onboarding` and `@metamask/forwarder`

The latest published versions of both libraries are now used. The
forwarder was added along with a convenience sript (`yarn forwarder`)
to make testing one-click onboarding easier.

Gudahtt commented:
This should be ready for review again. One of the signature-request tests added recently required updating due to the addition of a 'Connect' button on the `contract-test` dapp.

whymarrh commented on test/e2e/contract-test/index.html:
```diff
@@ -1,51 +1,64 @@
 <head>
+  <meta charset="UTF-8">
   <title>E2E Test Dapp</title>
+  <script src="node_modules/metamask-onboarding/dist/metamask-onboarding.bundle.js"></script>
```
Does this need to be updated with the new URL?

```suggestion
  <script src="node_modules/@metamask/onboarding/dist/metamask-onboarding.bundle.js"></script>
```

Gudahtt replied:
Yes, it does! Good catch.

Gudahtt replied:
This has been fixed

whymarrh commented:
I think this looks good, pending the one comment I left above. I think it might be useful to have either @tmashuang or someone to go through the flow to get an extra set of eyes on it before we :shipit: to the public.

Gudahtt commented:
Updated test instructions:

• Run `yarn dapp-forwarder` to start the test dapp and the forwarder simultaneously
• Navigate to `http://localhost:8080/` in a browser _without_ MetaMask installed (the button under the "Connect" header should say "Click here to install MetaMask!")
• Click the install button
• Naviate to chrome://extensions, enable dev mode, and manually install the MetaMask build
• Complete onboarding as usual, and you should see the button at the end with the updated notice
• Click the final button to be moved back to `http://localhost:8080/`

tmashuang reviewed (approved):
LGTM

whymarrh reviewed (approved):
LGTM
